### PR TITLE
TINKERPOP-1293: Implement GraphFilterStrategy as a default registration for GraphComputer

### PR DESCRIPTION
**Pull request by okram**
https://issues.apache.org/jira/browse/TINKERPOP-1293

`GraphFilterStrategy` (traversal optimization strategy) will create a graph filter for the OLAP traversal if one has not already been provided. This is made possible by `GraphFilterStrategy` analyzing the traversal to determine which edge directions/labels are being accessed and generating a traversal that meets that criteria. Along with this, `GraphFilter` has been advanced more intelligence, more test cases, more JavaDoc, and more helper methods for `GraphComputer` providers.

`GraphFilterStrategy` is a default `GraphComputer` strategy that is deactivated by `TinkerGraphComputer` as edge filters are expensive in `TinkerGraphComputer` given that the entire graph is already loaded into memory and any filtering requires labeled edges as being filtered (or not).

`mvn clean install`, Spark integration tests passing, and Giraph integration tests passing thus far...

VOTE +1.

### CODE REVIEW

**Comment by okram** [comment]
CHANGELOG

```
* `GraphFilter` helper methods are now more intelligent when determining edge direction/label legality.
* Added `GraphFilterStrategy` to automatically construct `GraphFilters` via traversal introspection.
```

UPDATE

```
GraphFilter and GraphFilterStrategy
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

`GraphFilter` has been significantly advanced where the determination of an edge direction/label legality is more stringent. Along with this, `GraphFilter.getLegallyPositiveEdgeLabels()` has been added as a helper method to make it easier for providers to know the space of labels being accessed by the traversal.

Note that `GraphFilterStrategy` is a default `TraversalStrategy` registered with `GraphComputer.` If `GraphFilter` is expensive for the underlying `GraphComputer` implementation, it can be deactivated accordingly as is done for `TinkerGraphComputer`.

[source,java]
----
static {
  TraversalStrategies.GlobalCache.registerStrategies(TinkerGraphComputer.class,
    TraversalStrategies.GlobalCache.getStrategies(GraphComputer.class).clone().removeStrategies(GraphFilterStrategy.class));
}
----
```

**Comment by okram** [comment]
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Apache TinkerPop .................................. SUCCESS [5.532s]
[INFO] Apache TinkerPop :: Gremlin Shaded ................ SUCCESS [2.355s]
[INFO] Apache TinkerPop :: Gremlin Core .................. SUCCESS [44.087s]
[INFO] Apache TinkerPop :: Gremlin Test .................. SUCCESS [11.803s]
[INFO] Apache TinkerPop :: Gremlin Groovy ................ SUCCESS [40.912s]
[INFO] Apache TinkerPop :: Gremlin Groovy Test ........... SUCCESS [7.801s]
[INFO] Apache TinkerPop :: TinkerGraph Gremlin ........... SUCCESS [3:29.140s]
[INFO] Apache TinkerPop :: Gremlin Benchmark ............. SUCCESS [6.440s]
[INFO] Apache TinkerPop :: Hadoop Gremlin ................ SUCCESS [11:07.210s]
[INFO] Apache TinkerPop :: Spark Gremlin ................. SUCCESS [6:18.138s]
[INFO] Apache TinkerPop :: Giraph Gremlin ................ SUCCESS [2:43:57.096s]
[INFO] Apache TinkerPop :: Neo4j Gremlin ................. SUCCESS [1:09:07.962s]
[INFO] Apache TinkerPop :: Gremlin Driver ................ SUCCESS [12.497s]
[INFO] Apache TinkerPop :: Gremlin Server ................ SUCCESS [13:22.778s]
[INFO] Apache TinkerPop :: Gremlin Console ............... SUCCESS [1:26.498s]
[INFO] Apache TinkerPop :: Gremlin Archetype ............. SUCCESS [0.161s]
[INFO] Apache TinkerPop :: Archetype - TinkerGraph ....... SUCCESS [5.221s]
[INFO] Apache TinkerPop :: Archetype - Server ............ SUCCESS [10.874s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 4:31:17.006s
[INFO] Finished at: Tue May 10 17:55:05 MDT 2016
[INFO] Final Memory: 105M/796M
[INFO] ------------------------------------------------------------------------
```

**Comment by dkuppitz** [comment]
VOTE: +1

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/process/computer/TinkerGraphComputer.java [hunk 1]
```diff
@@ -54,6 +56,12 @@
  */
 public final class TinkerGraphComputer implements GraphComputer {
 
+    static {
+        // GraphFilters are expensive w/ TinkerGraphComputer as everything is already in memory
```
Comment (analytically) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/process/computer/TinkerGraphComputer.java]: Would it be possible to explain this a bit more?

Comment (okram) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/process/computer/TinkerGraphComputer.java]: `TinkerGraphComputer` already has the entire graph loaded in memory. Thus, when you add a `GraphFilter`, what I'm doing is creating a `Set<Object>` of edge IDs that are allowed. Then, every `Vertex.edges()` is filtered by `edge.id() in Set<Object>`. Its lame. As such, edge-based `GraphFilters` are expensive in `TinkerGraphComputer`.

**Comment by spmallette** [comment]
Nice looking tests. ran `mvn clean install` - all good

VOTE +1

**Pull request merged**